File: NeuralNetNative/RbmTrainMethod.cpp
#define NEURALNETNATIVEAPI

#include <mathimf.h>
#include "RbmTrainMethod.h"
#include <cfloat>

namespace NeuralNetNative {
	namespace RestrictedBoltzmannMachine {
        RbmTrainMethod::RbmTrainMethod(StandardTypesNative::TrainSingle **trainData,
                       int trainDataSize,
                       GradientFunction *gradientFunction) {
            _trainDataIterator = new StandardTypesNative::RandomAccessIterator<StandardTypesNative::
                                                          TrainSingle*>(trainData, trainDataSize);
            _neuronNetOutput = (float*)_mm_malloc(trainData[0]->InputLength()*sizeof(float), 32);
			_testData = 0;
			_testDataSize = 0;

            _gradientFunction = gradientFunction;
        }

        RbmTrainMethod::RbmTrainMethod(StandardTypesNative::TrainSingle **trainData,
                       StandardTypesNative::TrainSingle **testData,
                       int trainDataSize, int testDataSize,
                       GradientFunction *gradientFunction) {
            _trainDataIterator = new StandardTypesNative::RandomAccessIterator<StandardTypesNative::
                                                          TrainSingle*>(trainData, trainDataSize);
            _neuronNetOutput = (float*)_mm_malloc(trainData[0]->InputLength()*sizeof(float), 32);
			_testData = testData;
			_testDataSize = testDataSize;

            _gradientFunction = gradientFunction;
        }

        RbmTrainMethod::~RbmTrainMethod(void) {
            _mm_free(_neuronNetOutput);
            
            delete _trainDataIterator;
            
            if (gradients != 0) {
                delete gradients;
                gradients = 0;
            }
        }

        void RbmTrainMethod::RunIterativeProcess(void) {
            if (IsTestDataAvailable()) {
	            RunTraingWithTesting();
	        }
			else {
				RunTraingWithoutTesting();
			}
        }

        void RbmTrainMethod::ApplyResults(void) {
            if (ProcessSate == StandardTypesNative::IterativeProcessState::Finished) {
				DeleteTemporaryData();

                if (gradients != 0) {
                    delete gradients;
                    gradients = 0;
                }
			}
        }

        bool RbmTrainMethod::IsTestDataAvailable() const {
            return !(_testData == 0 || _testDataSize == 0);
        }

        void RbmTrainMethod::RunTraingWithTesting(void) {
            float trainError = TestModel(_trainDataIterator->Collection(), _trainDataIterator->Size());
			float slidingTestError = TestModel(_testData, _testDataSize);
			float minTestError = slidingTestError;
			epochNumber = 1;
			while ((ProcessSate == StandardTypesNative::IterativeProcessState::InProgress) && 
				(trainError > properties->Epsilon) && 
				(epochNumber <= properties->MaxIterationCount) &&
				((epochNumber <= properties->SkipCvLimitFirstIterations) ||
                 (fabsf(slidingTestError - minTestError) < properties->CvLimit))) {

				TrainEpoch();

				trainError = TestModel(_trainDataIterator->Collection(), _trainDataIterator->Size());
				float testError = TestModel(_testData, _testDataSize);
                slidingTestError = properties->CvSlidingFactor*testError +
					(1.0f - properties->CvSlidingFactor)*slidingTestError;

				if (testError < minTestError) {
					minTestError = testError;
				}

				OnIterationCompleted(epochNumber, trainError, testError);
				epochNumber++;
			}
			OnIterativeProcessFinished(epochNumber);
        }

        void RbmTrainMethod::RunTraingWithoutTesting(void) {
            float trainError = TestModel(_trainDataIterator->Collection(), _trainDataIterator->Size());
			epochNumber = 1;
			while ((ProcessSate == StandardTypesNative::IterativeProcessState::InProgress) && 
				   (trainError > properties->Epsilon) && 
				   (epochNumber <= properties->MaxIterationCount)) {

				TrainEpoch();

				trainError = TestModel(_trainDataIterator->Collection(), _trainDataIterator->Size());

				OnIterationCompleted(epochNumber, trainError, std::numeric_limits<float>::quiet_NaN());
				epochNumber++;
			}
			OnIterativeProcessFinished(epochNumber);
        }

        int RbmTrainMethod::CalculatePackagesCount(void) const {
            int count = _trainDataIterator->Size()/properties->PackageSize;
			if (_trainDataIterator->Size()%properties->PackageSize != 0) {
				count++;
			}
			return count;
        }

        float RbmTrainMethod::TestModel(StandardTypesNative::TrainSingle **data, int dataSize) const {
            float sumError = 0.0f;
            for (int i = 0; i < dataSize; i++) {
                StandardTypesNative::TrainSingle *testExample = data[i];
                neuralNet->Predict(testExample->Input(), _neuronNetOutput);
                sumError += properties->Metrics->Calculate(testExample->Input(), _neuronNetOutput, visibleStatesCount);
            }
            return sumError / dataSize;
        }

        void RbmTrainMethod::TrainEpoch(void) {
            _trainDataIterator->RefreshRandomAccess();
			for (int i = 0; i < packagesCount; i++) {
				TrainPackage(i);
			}
        }

        void RbmTrainMethod::TrainPackage(int packageId) {
            _gradientFunction->PrepareToNextPackage(properties->PackageSize);
			for (int i = 0; i < properties->PackageSize; i++) {
				float *input = _trainDataIterator->Next()->Input();

				MakePositivePhase(input);
                _gradientFunction->StorePositivePhaseData(input, neuralNet->GetHiddenStates());
				MakeNegativePhase(packageId);
				_gradientFunction->StoreNegativePhaseData(GetVisibleStatesOnNegativePhase(packageId), GetHiddenStatesOnNegativePhase());
				RestoreVisibleStates(packageId);
			}
			_gradientFunction->MakeGradient(_packageFactor);
			ModifyWeightsOfNeuronNet();
        }

        void RbmTrainMethod::InitilazeMethod(NeuralNet *newNeuralNet, TrainProperties *newProperties) {
			neuralNet = dynamic_cast<RestrictedBoltzmannMachineBase*>(newNeuralNet);
			if (neuralNet == 0) {
				return;
			}

            visibleStatesCount = neuralNet->GetVisibleStatesCount();
            hiddenStatesCount = neuralNet->GetHiddenStatesCount();

            gradients = new RbmGradients(visibleStatesCount, hiddenStatesCount);
			_gradientFunction->Initialize(gradients);
				
			properties = newProperties;
			_packageFactor = 1.0f/properties->PackageSize;
			packagesCount = CalculatePackagesCount();

			CreateTemporaryData();

			ProcessSate = StandardTypesNative::IterativeProcessState::NotStarted;
		}

		TrainProperties* RbmTrainMethod::Properties(void) const {
			return properties;
		}
	}
}

File: NeuralNet/NeuralNets/RestrictedBoltzmannMachine/GenerativeRbm/TrainMethods/RbmTrainMethod.cs
using System;
using System.Collections.Generic;
using StandardTypes;

namespace NeuralNet.GenerativeRbm {
	public abstract class RbmTrainMethod : TrainMethod<TrainSingle> {
		private readonly RandomAccessIterator<TrainSingle> _trainDataIterator;
		private readonly IList<TrainSingle> _testData;
		private readonly IGradientFunction _gradientFunction;
		private float[] _neuronNetOutput;
        private float packageFactor;
		protected RbmGradients gradients;
		protected ITrainProperties<TrainSingle> properties;
		protected RestrictedBoltzmannMachine neuralNet;
		protected int visibleStatesCount;
		protected int hiddenStatesCount;
		protected int epochNumber;
		protected int packagesCount;

		protected RbmTrainMethod(IList<TrainSingle> trainData, IGradientFunction gradient) {
			_gradientFunction = gradient;
			_trainDataIterator = new RandomAccessIterator<TrainSingle>(trainData);
		}

		protected RbmTrainMethod(IList<TrainSingle> trainData, IList<TrainSingle> testData, IGradientFunction gradient) {
			_trainDataIterator = new RandomAccessIterator<TrainSingle>(trainData);
			_testData = testData;
			_gradientFunction = gradient;
		}

		public override void InitilazeMethod(INeuralNet neuralNet, ITrainProperties<TrainSingle> trainProperties) {
			if (!(neuralNet is RestrictedBoltzmannMachine)) {
				throw new ArgumentException("Neural net has other structure");
			}

			this.neuralNet = (RestrictedBoltzmannMachine) neuralNet;
			visibleStatesCount = this.neuralNet.VisibleStates.Length;
			hiddenStatesCount = this.neuralNet.HiddenStates.Length;

			gradients = new RbmGradients(visibleStatesCount, hiddenStatesCount);
			_gradientFunction.Initialize(gradients);

			_neuronNetOutput = new float[visibleStatesCount];

			properties = trainProperties;
			packageFactor = 1.0f/properties.PackageSize;
			packagesCount = CalculatePackagesCount();

			AllocateMemory();

			ProcessSate = IterativeProcessState.NotStarted;
		}

		protected override void RunIterativeProcess() {
			if (IsTestDataAvailable()) {
	            RunTraingWithTesting();
	        }
			else {
				RunTraingWithoutTesting();
			}
		}

		protected override void ApplyResults() {
			if (ProcessSate == IterativeProcessState.Finished) {
				ClearReference();
			}
		}

		public override ITrainProperties<TrainSingle> Properties {
			get { return properties; }
		}

		protected abstract void AllocateMemory();

		protected abstract void ClearReference();

		private bool IsTestDataAvailable() {
			return !(_testData == null || _testData.Count == 0);
		}

		private void RunTraingWithTesting() {
			var trainError = TestModel(_trainDataIterator.Collection);
			var slidingTestError = TestModel(_testData);
			var minTestError = slidingTestError;
			epochNumber = 1;
			while ((ProcessSate == IterativeProcessState.InProgress) &&
				   (trainError > properties.Epsilon) &&
				   (epochNumber <= properties.MaxIterationCount) &&
				   ((epochNumber <= properties.SkipCvLimitFirstIterations) ||
				    (Math.Abs(slidingTestError - minTestError) < properties.CvLimit))) {
				
				TrainEpoch();

				trainError = TestModel(_trainDataIterator.Collection);
			    var testError = TestModel(_testData);
				slidingTestError = properties.CvSlidingFactor*testError +
					(1f - properties.CvSlidingFactor)*slidingTestError;
				
				if (testError < minTestError) {
					minTestError = testError;
				}

                OnIterationCompleted(new IterationCompletedEventArgs(epochNumber, trainError, testError));
				epochNumber++;
			}
			OnIterativeProcessFinished(new IterativeProcessFinishedEventArgs(epochNumber));
		}

		private void RunTraingWithoutTesting() {
			var trainError = TestModel(_trainDataIterator.Collection);
			epochNumber = 1;
			while ((ProcessSate == IterativeProcessState.InProgress) && 
				   (trainError > properties.Epsilon) && 
				   (epochNumber <= properties.MaxIterationCount)) {
				
				TrainEpoch();

				trainError = TestModel(_trainDataIterator.Collection);

                OnIterationCompleted(new IterationCompletedEventArgs(epochNumber, trainError, float.NaN));
				epochNumber++;
			}
			OnIterativeProcessFinished(new IterativeProcessFinishedEventArgs(epochNumber));
		}

		private int CalculatePackagesCount() {
			var count = _trainDataIterator.Size()/properties.PackageSize;
			if (_trainDataIterator.Size()%properties.PackageSize != 0) {
				count++;
			}
			return count;
		}

		private float TestModel(IList<TrainSingle> data) {
			var sumError = 0.0f;
			for (var i = 0; i < data.Count; i++) {
				var testExample = data[i];
				neuralNet.Predict(testExample.Input, _neuronNetOutput);
				sumError += properties.Metrics.Calculate(testExample.Input, _neuronNetOutput);
			}
			return sumError / data.Count;
		}

		private void TrainEpoch() {
			_trainDataIterator.RefreshRandomAccess();
			for (var i = 0; i < packagesCount; i++) {
				TrainPackage(i);
			}
		}

		private void TrainPackage(int packageId) {
			_gradientFunction.PrepareToNextPackage(properties.PackageSize);
			for (var i = 0; i < properties.PackageSize; i++) {
				var input = _trainDataIterator.Next().Input;

				MakePositivePhase(input);
				_gradientFunction.StorePositivePhaseData(input, neuralNet.HiddenStates);
				MakeNegativePhase(packageId);
				_gradientFunction.StoreNegativePhaseData(GetVisibleStatesOnNegativePhase(packageId), GetHiddenStatesOnNegativePhase());
				RestoreVisibleStates(packageId);
			}
			_gradientFunction.MakeGradient(packageFactor);
			ModifyWeightsOfNeuronNet();
		}

		protected abstract void MakePositivePhase(float[] input);

		protected abstract void MakeNegativePhase(int packageId);

		protected abstract float[] GetVisibleStatesOnNegativePhase(int packageId);

		protected abstract float[] GetHiddenStatesOnNegativePhase();

		protected abstract void RestoreVisibleStates(int packageId);

		protected abstract void ModifyWeightsOfNeuronNet();
	}
}
